libr/anal/block.c

```diff
@@ -0,0 +1,579 @@
+					break;
+				}
+			}
+			if (!r_list_contains (fcn->bbs, block)) {
+				eprintf ("FUCK: Fcn %s is referenced by block @ 0x%"PFMT64x", but block is not referenced by function\n", fcn->name, block->addr);
+			}
+		}
+#endif
+	}
+
+#if DEEPERCHECKS
+	RListIter *fcniter;
+	RAnalFunction *fcn;
+	r_list_foreach (anal->fcns, fcniter, fcn) {
+		RListIter *blockiter;
+		ut64 min = UT64_MAX;
+		ut64 max = UT64_MIN;
+		ut64 realsz = 0;
+		r_list_foreach (fcn->bbs, blockiter, block) {
+			RListIter *blockiter2;
+			RAnalBlock *block2;
+			if (block->addr < min) {
+				min = block->addr;
+			}
+			if (block->addr + block->size > max) {
+				max = block->addr + block->size;
+			}
+			realsz += block->size;
+			for (blockiter2 = blockiter->n; blockiter2 && (block2 = blockiter2->data, 1); blockiter2 = blockiter2->n) {
+				if (block == block2) {
+					eprintf("FUCK: Duplicate basic block @ 0x%"PFMT64x" in function %s\n", block->addr, fcn->name);
+					break;
+				}
+			}
+			if (!r_list_contains (block->fcns, fcn)) {
+				eprintf("FUCK: block @ 0x%"PFMT64x" is referenced by Fcn %s, not the other way around\n", block->addr, fcn->name);
+			}
+		}
+
+		if (fcn->meta._min != UT64_MAX && (fcn->meta._min != min || fcn->meta._max != max)) {
+			eprintf("SHIP: min/max wrong!!!!!\n");
+		}
+		if (r_anal_fcn_realsize (fcn) != realsz) {
+			eprintf("SHIP: realsize wrong!!!!!!\n");
+		}
+	}
+#endif
+#endif
+}
+
+// TODO: this can be removed just before merging too
+R_API void r_anal_block_check_leaks(RAnal *anal) {
+	RBIter iter;
+	RAnalBlock *block;
+	r_rbtree_foreach (anal->bb_tree, iter, block, RAnalBlock, _rb) {
+		if (block->ref != r_list_length (block->fcns))  {
+			eprintf ("LEEEEEEEEEEEEEAK!!! bb: 0x%"PFMT64x"\n", block->addr);
+		}
+	}
+}
+
+void __block_free_rb(RBNode *node, void *user) {
+	RAnalBlock *block = unwrap (node);
+	block_free (block);
+}
+
+R_API RAnalBlock *r_anal_block_get_at(RAnal *anal, ut64 addr) {
+	RBNode *node = r_rbtree_find (anal->bb_tree, &addr, __bb_addr_cmp, NULL);
+	if (!node) {
+		return NULL;
+	}
+	return unwrap (node);
+}
+
+// This is a special case of what r_interval_node_all_in() does
+static bool all_in(RAnalBlock *node, ut64 addr, RAnalBlockCb cb, void *user) {
+	while (node && addr < node->addr) {
+		// less than the current node, but might still be contained further down
+		node = unwrap (node->_rb.child[0]);
+	}
+	if (!node) {
+		return true;
+	}
+	if (addr >= node->_max_end) {
+		return true;
+	}
+	if (addr < node->addr + node->size) {
+		if (!cb (node, user)) {
+			return false;
+		}
+	}
+	// This can be done more efficiently by building the stack manually
+	if (!all_in (unwrap (node->_rb.child[0]), addr, cb, user)) {
+		return false;
+	}
+	if (!all_in (unwrap (node->_rb.child[1]), addr, cb, user)) {
+		return false;
+	}
+	return true;
+}
+
+R_API bool r_anal_block_get_in(RAnal *anal, ut64 addr, RAnalBlockCb cb, void *user) {
```

Comment:
this name is misleading, because `get` seems like you get a AnalBlock in return. I think it makes more sense to use `r_anal_block_foreach_in`. WDYT?

Reply:
sgtm